File: src/main/scala/com/fillmore_labs/talk/value/simple/scala/EmailAddress.scala
package com.fillmore_labs.talk.value.simple.scala

import org.apache.commons.validator.routines.EmailValidator

object EmailAddress {
  private val Validator: EmailValidator = EmailValidator.getInstance()

  /** Java API: Creates an e-mail address.
    *
    * @param address An address.
    * @return An e-mail address.
    */
  def of(address: String): EmailAddress = EmailAddress(address)
}

final case class EmailAddress private (address: String) {
  require(
    EmailAddress.Validator.isValid(address),
    s"$address is not a valid e-mail address"
  )

  /** Java API: The given name of this person. Everyone has one.
    *
    * @return An address.
    */
  def getAddress: String = address
}

File: src/main/kotlin/com/fillmore_labs/talk/value/simple/kotlin/EmailAddress.kt
package com.fillmore_labs.talk.value.simple.kotlin

import org.apache.commons.validator.routines.EmailValidator

private val EMAIL_VALIDATOR: EmailValidator = EmailValidator.getInstance()

/**
 * A (validated) e-mail address.
 *
 * @param[address] Set the e-mail address.
 */
@Suppress("DataClassPrivateConstructor") // https://youtrack.jetbrains.com/issue/KT-11914
data class EmailAddress private constructor(val address: String) {
    init {
        require(EMAIL_VALIDATOR.isValid(address)) { "$address is not a valid e-mail address" }
    }

    companion object {
        @JvmStatic
        fun of(address: String) = EmailAddress(address)
    }
}
